examples/oft-solana/tasks/solana/index.ts

```
@@ -42,29 +41,99 @@ const LOOKUP_TABLE_ADDRESS: Partial<Record<EndpointId, PublicKey>> = {
 }
 
-const getFromEnv = (key: string): string => {
+/**
+ * Extracts the value of the given environment variable.
+ * @todo consider removing this since it's no longer used (replaced by using @solana-developers/helpers)
+ * @param key The name of the environment variable to extract.
+ * @param optional Whether the environment variable is optional or not.
+ * If it is, the function will return undefined if the variable is not defined.
+ * Otherwise, it will throw an error if the variable is not defined.
+ * @returns The value of the environment variable, or undefined if optional and not defined.
+ */
+const getFromEnv = (key: string, optional = false): string | undefined => {
     const value = process.env[key]
-    if (!value) {
+    if (!value && !optional) {
         throw new Error(`${key} is not defined in the environment variables.`)
     }
     return value
 }
 
-/**
- * Extracts the SOLANA_PRIVATE_KEY from the environment.  This is purposely not exported for encapsulation purposes.
- */
-const getSolanaPrivateKeyFromEnv = () => getFromEnv('SOLANA_PRIVATE_KEY')
+// TODO in another PR: considerg moving keypair related functions to tasks/solana/utils.ts
+async function getSolanaKeypair(readOnly = false): Promise<Keypair> {
+    const logger = createLogger()
+
+    // Early exit if read-only: ephemeral Keypair is enough.
+    if (readOnly) {
+        logger.info('Read-only mode: Using ephemeral (randomly generated) keypair.')
+        return Keypair.generate()
+    }
+
+    // Attempt to load from each source
+    const keypairEnvPrivate = process.env.SOLANA_PRIVATE_KEY
+        ? getKeypairFromEnvironment('SOLANA_PRIVATE_KEY')
+        : undefined // #1 SOLANA_PRIVATE_KEY
+    const keypairEnvPath = process.env.SOLANA_KEYPAIR_PATH
+        ? await getKeypairFromFile(process.env.SOLANA_KEYPAIR_PATH)
+        : undefined // #2 SOLANA_KEYPAIR_PATH
+    const keypairDefaultPath = await getKeypairFromFile() // #3 ~/.config/solana/id.json
```

This should be in a try/catch.  If the file does not exist or is not readable, it will throw:

https://github.com/solana-developers/helpers/blob/main/src/lib/keypair.ts#L33

---

I think we should just let the exception bubble up, as the functions already handle various invalid situations:

e.g. https://github.com/solana-developers/helpers/blob/main/src/lib/keypair.ts#L33-L45
````
  } catch (error) {
    throw new Error(`Could not read keypair from file at '${filepath}'`);
  }

  // Parse contents of file
  let parsedFileContents: Uint8Array;
  try {
    parsedFileContents = Uint8Array.from(JSON.parse(fileContents));
  } catch (thrownObject) {
    const error = thrownObject as Error;
    if (!error.message.includes("Unexpected token")) {
      throw error;
    }
    throw new Error(`Invalid secret key file at '${filepath}'!`);
  }
  return Keypair.fromSecretKey(parsedFileContents);
};
```

if we were to catch it, we either will lose the context already given for free, or we'd need to do error message matching again, duplicating the code.

---

or, if you think it's better, then it'll just be catching to add additional context to it, e.g.

```typescript
let keypairEnvPath

try {
  keypairEnvPath = process.env.SOLANA_KEYPAIR_PATH
        ? await getKeypairFromFile(process.env.SOLANA_KEYPAIR_PATH)
        : undefined // #2 SOLANA_KEYPAIR_PATH
} catch (e) {
throw new Error(`Error when trying keypairEnvPath provided via process.env.SOLANA_KEYPAIR_PATH: ${e.message}`)
}

```

---

My point is that not everyone has a key defined at ~/.config/solana/id.json.  If we assume it is there, it could throw an exception if someone is using the environment variables instead.

---

You're right. I overlooked that case. let me update the PR to handle this.

---

@ryandgoulding I've pushed the fix. thanks again for spotting the issue.